src/irmin-test/layered_store.ml

```diff
@@ -0,0 +1,958 @@
+            `Upper)
+        [ 0; 1; 3 ]
+      >>= fun () ->
+      Lwt.catch
+        (fun () ->
+          S.layer_id repo (S.Content_t (List.nth contents 2)) >|= fun _ ->
+          Alcotest.fail "Should not find x2")
+        (function Not_found -> Lwt.return_unit | exn -> Lwt.fail exn)
+      >>= fun () ->
+      Lwt.catch
+        (fun () ->
+          S.layer_id repo (S.Content_t (List.nth contents 4)) >|= fun _ ->
+          Alcotest.fail "Should not found x4")
+        (function Not_found -> Lwt.return_unit | exn -> Lwt.fail exn)
+      >>= fun () ->
+      Lwt_list.iter_p
+        (fun i ->
+          check_layer_id repo
+            (S.Content_t (List.nth contents i))
+            ("layer id of x" ^ string_of_int i)
+            `Upper)
+        [ 3; 5; 6 ]
+      >>= fun () -> S.Repo.close repo
+    in
+    run x test
+
+  module Hook = S.PrivateLayer.Hook
+
+  let before_copy f =
+    let p, r = Lwt.wait () in
+    let hook =
+      Hook.v (function
+        | `Before_Copy ->
+            Log.debug (fun l -> l "test hook -- before_copy");
+            Lwt.async (fun () -> f () >|= fun () -> Lwt.wakeup r ());
+            Lwt.return_unit
+        | _ -> Lwt.return_unit)
+    in
+    (hook, p)
+
+  let before_copy_newies f =
+    let p, r = Lwt.wait () in
+    let hook =
+      Hook.v (function
+        | `Before_Copy_Newies ->
+            Log.debug (fun l -> l "test hook -- before_copy_newies");
+            Lwt.async (fun () -> f () >|= fun () -> Lwt.wakeup r ());
+            Lwt.return_unit
+        | _ -> Lwt.return_unit)
+    in
+    (hook, p)
+
+  let before_copy_last_newies f =
+    let p, r = Lwt.wait () in
+    let hook =
+      Hook.v (function
+        | `Before_Copy_Last_Newies ->
+            Log.debug (fun l -> l "test hook -- before_copy_last_newies");
+            Lwt.async (fun () -> f () >|= fun () -> Lwt.wakeup r ());
+            Lwt.return_unit
+        | _ -> Lwt.return_unit)
+    in
+    (hook, p)
+
+  let before_flip f =
+    let p, r = Lwt.wait () in
+    let hook =
+      Hook.v (function
+        | `Before_Flip ->
+            Log.debug (fun l -> l "test hook -- before_flip");
+            Lwt.async (fun () -> f () >|= fun () -> Lwt.wakeup r ());
+            Lwt.return_unit
+        | _ -> Lwt.return_unit)
+    in
+    (hook, p)
+
+  let before_clear f =
+    let p, r = Lwt.wait () in
+    let hook =
+      Hook.v (function
+        | `Before_Clear ->
+            Log.debug (fun l -> l "test hook -- before_clear");
+            Lwt.async (fun () -> f () >|= fun () -> Lwt.wakeup r ());
+            Lwt.return_unit
+        | _ -> Lwt.return_unit)
+    in
+    (hook, p)
+
+  let after_clear f =
+    let p, r = Lwt.wait () in
+    let hook =
+      Hook.v (function
+        | `After_Clear ->
+            Log.debug (fun l -> l "test hook -- after_clear");
+            Lwt.async (fun () -> f () >|= fun () -> Lwt.wakeup r ());
+            Lwt.return_unit
+        | _ -> Lwt.return_unit)
+    in
+    (hook, p)
```

Comment:
These look like they could be cut down with a helper function :slightly_smiling_face: